File: frontend/src/api/music.ts
import request from './axios';
import type {Artist, Album, Music, MusicListResponse, InteractionCreate, InteractionResponse, LikeStatusResponse} from '@/types/entity';

export const musicApi = {
    // 获取艺术家列表
    getArtists: () => request.get<any, Artist[]>('/music/artists'),

    // 创建艺术家
    createArtist: (data: { name: string; bio?: string }) =>
        request.post<any, Artist>('/music/artists', data),

    // 获取专辑列表
    getAlbumsByArtist: (artistId: number) =>
        request.get<any, Album[]>(`/music/artists/${artistId}/albums`),

    // 创建专辑
    createAlbum: (data: { title: string; release_date: string; artist_id: number }) =>
        request.post<any, Album>('/music/albums', data),

    // 上传音乐 (Multipart)
    uploadMusic: (file: File, meta: { title: string; duration: number; album_id: number; track_number?: number }) => {
        const formData = new FormData();
        formData.append('file', file);
        formData.append('title', meta.title);
        formData.append('duration', meta.duration.toString());
        formData.append('album_id', meta.album_id.toString());
        if (meta.track_number) {
            formData.append('track_number', meta.track_number.toString());
        }

        return request.post<any, Music>('/music/upload', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
    },

    // 获取音乐列表
    getMusicList: (skip = 0, limit = 50) =>
        request.get<any, MusicListResponse>('/music/', {params: {skip, limit}}),

    // 删除音乐
    deleteMusic: (id: number) =>
        request.delete<any, void>(`/music/${id}`),

    /**
     * 记录用户交互行为
     * 用于前端播放器事件上报
     *
     * @param data 交互数据，包含音乐ID和交互类型
     * @returns 创建的交互记录
     */
    recordInteraction: (data: InteractionCreate) =>
        request.post<any, InteractionResponse>('/music/interactions', data),

    /**
     * 检查用户是否已收藏某音乐
     * 用于展示收藏状态
     *
     * @param musicId 音乐ID
     * @returns 收藏状态
     */
    checkLikeStatus: (musicId: number) =>
        request.get<any, LikeStatusResponse>(`/music/interactions/like-status/${musicId}`),
};

File: backend/app/api/v1/endpoints/music.py
"""
音乐资源管理 API

包含艺术家、专辑的 CRUD 和音乐文件的上传接口。
"""

from typing import Annotated, List

from fastapi import APIRouter, Depends, File, Form, UploadFile
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.api import deps
from app.models.music import Album, Music  # 确保导入了模型
from app.models.user import User
from app.repositories.music_repository import AlbumRepository, ArtistRepository, MusicRepository
from app.schemas.music import (
    AlbumCreate, AlbumResponse,
    ArtistCreate, ArtistResponse,
    MusicCreate, MusicResponse,
    MusicListResponse
)
from app.services.music_service import music_service

router = APIRouter()


# --- Artist Endpoints ---

@router.post("/artists", response_model=ArtistResponse)
async def create_artist(
        *,
        db: Annotated[AsyncSession, Depends(deps.get_db)],
        artist_in: ArtistCreate,
        current_user: Annotated[User, Depends(deps.get_current_active_superuser)],
):
    """创建艺术家 (仅管理员)"""
    repo = ArtistRepository()
    return await repo.create(db, obj_in=artist_in)


@router.get("/artists", response_model=List[ArtistResponse])
async def read_artists(
        db: Annotated[AsyncSession, Depends(deps.get_db)],
        skip: int = 0,
        limit: int = 100,
):
    """获取艺术家列表 (公开)"""
    repo = ArtistRepository()
    return await repo.get_all(db, skip=skip, limit=limit)


# --- Album Endpoints ---

@router.post("/albums", response_model=AlbumResponse)
async def create_album(
        *,
        db: Annotated[AsyncSession, Depends(deps.get_db)],
        album_in: AlbumCreate,
        current_user: Annotated[User, Depends(deps.get_current_active_superuser)],
):
    """创建专辑 (仅管理员)"""
    repo = AlbumRepository()
    new_album = await repo.create(db, obj_in=album_in)

    # 显式加载关联的 artist，解决 MissingGreenlet 错误
    # 因为 Pydantic 响应模型需要嵌套的 artist 信息
    stmt = select(Album).options(selectinload(Album.artist)).where(Album.id == new_album.id)
    result = await db.execute(stmt)
    return result.scalar_one()


@router.get("/artists/{artist_id}/albums", response_model=List[AlbumResponse])
async def read_artist_albums(
        artist_id: int,
        db: Annotated[AsyncSession, Depends(deps.get_db)],
):
    """获取某艺术家的所有专辑"""
    repo = AlbumRepository()
    return await repo.get_by_artist(db, artist_id=artist_id)


# --- Music Endpoints ---

@router.post("/upload", response_model=MusicResponse)
async def upload_music(
        *,
        db: Annotated[AsyncSession, Depends(deps.get_db)],
        current_user: Annotated[User, Depends(deps.get_current_active_superuser)],
        file: Annotated[UploadFile, File(...)],
        title: Annotated[str, Form(...)],
        duration: Annotated[int, Form(...)],
        album_id: Annotated[int, Form(...)],
        track_number: Annotated[int, Form()] = 1,
):
    """
    上传音乐文件 (仅管理员)
    """
    # 构造 Schema 进行数据校验
    music_meta = MusicCreate(
        title=title,
        duration=duration,
        album_id=album_id,
        track_number=track_number
    )

    # 调用 Service 执行事务
    new_music = await music_service.upload_music(db, file, music_meta)

    # 显式加载关联的 album 和 album.artist
    # 因为 Pydantic 响应模型需要嵌套的 album 信息
    stmt = select(Music).options(
        selectinload(Music.album).selectinload(Album.artist)
    ).where(Music.id == new_music.id)
    result = await db.execute(stmt)
    return result.scalar_one()


@router.get("/", response_model=MusicListResponse)
async def read_musics(
        db: Annotated[AsyncSession, Depends(deps.get_db)],
        skip: int = 0,
        limit: int = 100,
):
    """获取音乐列表 (分页)"""
    # 获取总数
    count_stmt = select(func.count()).select_from(Music)
    count_result = await db.execute(count_stmt)
    total = count_result.scalar_one()

    # 列表查询也需要预加载，否则序列化时会报错
    # 这里的 repo.get_multi 默认没有预加载，我们需要重写一下查询
    stmt = select(Music).options(
        selectinload(Music.album).selectinload(Album.artist)
    ).offset(skip).limit(limit)
    result = await db.execute(stmt)
    items = list(result.scalars().all())

    return MusicListResponse(items=items, total=total)


# DELETE 接口
@router.delete("/{music_id}", status_code=204)
async def delete_music(
        music_id: int,
        db: Annotated[AsyncSession, Depends(deps.get_db)],
        # 权限校验: 仅管理员可删除
        current_user: Annotated[User, Depends(deps.get_current_active_superuser)],
):
    """
    删除音乐
    """
    await music_service.delete_music(db, music_id)
